streams/src/main/java/org/apache/kafka/streams/processor/internals/ReadOnlyTask.java

```diff
@@ -190,7 +190,7 @@ public void clearTaskTimeout() {
 
     @Override
     public boolean commitNeeded() {
-        throw new UnsupportedOperationException("This task is read-only");
+        return task.commitNeeded();
```

Comment:
Why do we need to change these two functions?

Reply:
It gives me a bit of a headache to forward the call to `commitNeeded()` since it changes the state of a stream task by updating the consumed offsets.
I am also wondering why Streams also commits restoring tasks.

Reply:
I think the actual culprit is like @lucasbru said, `StreamThread.maybeCommit uses TaskManager.allTasks`, while the callee `TaskManager.allTasks` is used in many different places.. For example for IQ, it was used and in that case it should return all tasks, whereas in this caller, `TaskManager.allTasks` should just return all processing tasks when updater is enabled, i.e. only the ones in `TaskRegistry`.

> I am also wondering why Streams also commits restoring tasks.

For that, the rationale is to allow the restoration progress to be recorded as well in case it was paused due to another rebalance (though it's `prepareCommit` would always return empty map, so the `committing` process for those restoring tasks would be reduced to writing checkpoints). But somehow in the middle of history we lost the logic to ever change `commitNeeded` flag for restoring tasks, so they would always not be triggered. Hence a new JIRA is proposed to fix it, and we fixed it in state updater.

I thought about fixing it in a probably better way, which involves more changes: 1) introduce a `TaskManager.allProcessingTasks`, 2) depending on the updater enabled flag, let `maybeCommit` call either this one or `allTasks`. When I thought about that, I admit I was thinking this change is a bit too much, and since tasks in updater the task should not make any side-effects and return `commitNeeded` false this maybe just okay.

But thinking about this a bit more, I think it's indeed worrisome to let `ReadyOnlyTask.commitNeeded` to have any potential side effects if we ever have bugs..